kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java

```diff
@@ -43,57 +51,68 @@ public KAppToTermConsVisitor(Module mod) {
                 return Left.apply(Sets.newHashSet(KEMException.innerParserError(msg, kl)));
-            } else if (prods.size() == 1)
-                return super.apply(TermCons.apply(items, prods.iterator().next(), tc.location(), tc.source()));
-            else {
-                // instantiate all labels found and let the type checker filter them out
-                Set<Term> tcs = new HashSet<>();
-                for (Production prd : prods)
-                    tcs.add(TermCons.apply(items, prd, tc.location(), tc.source()));
-                return super.apply(Ambiguity.apply(tcs, tc.location(), tc.source()));
-            }
+            } else if (sol.size() == 1) {
+                return super.apply(sol.iterator().next());
+            } else
+                return super.apply(Ambiguity.apply(sol, tc.location(), tc.source()));
         }
         return super.apply(tc);
     }
 
     /**  Recurse under #KList and flatten all the terms */
     private static PStack<Term> flattenKList(Term t) {
         if (t instanceof Ambiguity) {
-            Ambiguity amb = (Ambiguity) t;
-            // prefer KList if they exist, otherwise return t
-            List<PStack<Term>> klists = amb.items().stream()
-                    .filter(x -> x instanceof TermCons
-                            && ((TermCons) x).production().klabel().isDefined()
-                            && ((TermCons) x).production().klabel().get().name().equals("#KList"))
-                    .map(KAppToTermConsVisitor::flattenKList)
-                    .sorted((o1, o2) -> o2.size() - o1.size()).collect(Collectors.toList());
-
-            // expecting the first ambiguity branch to have the highest amount of elements
-            assert klists.size() <= 1 || klists.get(0).size() != klists.get(1).size():
-                    KAppToTermConsVisitor.class + ":" + " unexpected ambiguity pattern found under KApp " + amb;
-            if (klists.size() != 0)
-                return klists.get(0);
+            assert false : KAppToTermConsVisitor.class + " expected all ambiguities to already be pushed to the top:\n" +
+                    "   Source: " + ((Ambiguity) t).items().iterator().next().source().orElse(null) + "\n" +
+                    "   Location: " + ((Ambiguity) t).items().iterator().next().location().orElse(null);
         } else if (t instanceof TermCons) {
             TermCons tc = (TermCons) t;
-            if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#KList"))
-                return flattenKList(tc.items().get(1)).plusAll(flattenKList(tc.items().get(0)));
-            else if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#EmptyKList"))
+            if (tc.production().klabel().isDefined() && tc.production().klabel().get().equals(KLabels.KLIST))
+                return flattenKList(tc.get(0)).plusAll(flattenKList(tc.get(1)));
```

Comment:
Pretty sure this is reversed too

Reply:
I was wrong here too; plusAll prepends the argument in reverse order, so if I have ((1, 2),(3, 4)) as the children, I am going to end up with 1,2,4,3 as the final arguments.

Reply:
the second flattenKList call needs to be reversed.